### ISSUE TITLE
load data infile behavior differences between TiDB and MySQL when columns order is changed

### ISSUE THREAD

**Issue by yahonda** [pingcap/tidb#36302]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. Prepare `test.tsv` file that file needs separated by hard tab
```
1       100     test1
2       200     test2
3       300     \N
```
2. Start up `tiup playground. 

```
%  tiup playground
```

3. Login to TiDB using mysql cli with `--local-infile=1` option
```
% mysql --comments --host 127.0.0.1 --port 4000 -u root  --local-infile=1
```

4. At MySQL cli, set `SET GLOBAL local_infile=on; ` to allow load infile.
mysql> use test;
mysql> SET GLOBAL local_infile=on; 

5. Create table `test`

```
mysql> create table test (id int primary key, num int, str text);
```
6. Execute `load data local infile` those columns order is different from the table definition and test.tsv file.
```
mysql> load data local infile 'test.tsv' into table test (id, str, num);
```
7. Select test table

### 3. What did you see instead (Required)
- TiDB playground
```
mysql> select * from test;
+----+------+-------+
| id | num  | str   |
+----+------+-------+
|  1 |  100 | test1 |
|  2 |  200 | test2 |
|  3 |  300 | NULL  |
+----+------+-------+
3 rows in set (0.01 sec)

mysql>
```

- Result with MySQL 8.0.29
```
mysql> load data local infile 'test.tsv' into table test (id, str, num);
Query OK, 3 rows affected, 2 warnings (0.01 sec)
Records: 3  Deleted: 0  Skipped: 0  Warnings: 2

mysql> show warnings;
+---------+------+------------------------------------------------------------+
| Level   | Code | Message                                                    |
+---------+------+------------------------------------------------------------+
| Warning | 1366 | Incorrect integer value: 'test1' for column 'num' at row 1 |
| Warning | 1366 | Incorrect integer value: 'test2' for column 'num' at row 2 |
+---------+------+------------------------------------------------------------+
2 rows in set (0.00 sec)

mysql> select * from test;
+----+------+------+
| id | num  | str  |
+----+------+------+
|  1 |    0 | 100  |
|  2 |    0 | 200  |
|  3 | NULL | 300  |
+----+------+------+
3 rows in set (0.00 sec)

mysql> select version();
+-----------+
| version() |
+-----------+
| 8.0.29    |
+-----------+
1 row in set (0.00 sec)

mysql>
```

### 4. What is your TiDB version? (Required)

```
mysql> select tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v6.1.0
Edition: Community
Git Commit Hash: 1a89decdb192cbdce6a7b0020d71128bc964d30f
Git Branch: heads/refs/tags/v6.1.0
UTC Build Time: 2022-06-05 05:09:34
GoVersion: go1.18.2
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.00 sec)
```

**Comment by zyguan** [comment]
This issue shall be already fixed by #35222.

**Comment by yahonda** [comment]
Confirmed it has been fixed.

```sql
% mysql --comments --host 127.0.0.1 --port 4000 -u root  --local-infile=1
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 407
Server version: 5.7.25-TiDB-v6.4.0 TiDB Server (Apache License 2.0) Community Edition, MySQL 5.7 compatible

Copyright (c) 2000, 2022, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> use test;
Database changed
mysql> SET GLOBAL local_infile=on;
Query OK, 0 rows affected (0.03 sec)

mysql> create table test (id int primary key, num int, str text);
Query OK, 0 rows affected (0.11 sec)

mysql> load data local infile 'test.tsv' into table test (id, str, num);
Query OK, 3 rows affected, 2 warnings (0.01 sec)
Records: 3  Deleted: 0  Skipped: 0  Warnings: 2

mysql> show warnings;
+---------+------+-------------------------------------------+
| Level   | Code | Message                                   |
+---------+------+-------------------------------------------+
| Warning | 1292 | Truncated incorrect DOUBLE value: 'test1' |
| Warning | 1292 | Truncated incorrect DOUBLE value: 'test2' |
+---------+------+-------------------------------------------+
2 rows in set (0.00 sec)

mysql>  select * from test;
+----+------+------+
| id | num  | str  |
+----+------+------+
|  1 |    0 | 100  |
|  2 |    0 | 200  |
|  3 | NULL | 300  |
+----+------+------+
3 rows in set (0.01 sec)

mysql> select tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v6.4.0
Edition: Community
Git Commit Hash: cf36a9ce2fe1039db3cf3444d51930b887df18a1
Git Branch: heads/refs/tags/v6.4.0
UTC Build Time: 2022-11-13 05:15:26
GoVersion: go1.19.2
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: tikv
1 row in set (0.00 sec)

mysql>
```
